File: Chapter 03/JS/section_3_3_4_isp/after/goalie.js
/**
 * INTERFACE COMPOSITION (The Goalie "Opt-In").
 * * ARCHITECTURE NOTE: The beauty of Interface Segregation is that classes 
 * can "opt-in" to the specific behaviors they need. Since a Goalie needs 
 * both field skills (kicking/tackling) and specialized net skills, it 
 * simply implements both conceptual roles. 
 * * We have accommodated the complex requirements of the Goalie without 
 * polluting the Midfielder's codebase with irrelevant methods!
 */
class Goalie {
    /**
     * Implements Field Player logic.
     */
    practiceShooting() {
        console.log("  [Goalie] Practicing goal kicks and long shots.");
    }

    /**
     * Implements Field Player logic.
     */
    practiceTackling() {
        console.log("  [Goalie] Practicing 1-on-1 box tackles.");
    }

    /**
     * Implements Goalie-specific logic.
     */
    practiceDivingSaves() {
        console.log("  [Goalie] Practicing top-corner diving saves.");
    }

    /**
     * Implements Goalie-specific logic.
     */
    practiceHandDistribution() {
        console.log("  [Goalie] Practicing fast break throws.");
    }
}

module.exports = Goalie;

File: Chapter 03/C#/Section_3_3_4_ISP/After/Goalie.cs
using System;

namespace Chapter03.ISP.After;

/// <summary>
/// Interface Composition.
/// 
/// ARCHITECTURE NOTE: The beauty of Interface Segregation is that classes 
/// can just "opt-in" to the behaviors they actually need. Because the Goalie 
/// needs to practice everything, it simply implements BOTH interfaces. 
/// 
/// We accommodated the complex requirements of the Goalie without polluting 
/// the Midfielder's codebase!
/// </summary>
public class Goalie : IFieldPlayerTraining, IGoalieTraining
{
    public void PracticeShooting() 
    { 
        Console.WriteLine("  [Goalie] Practicing goal kicks and long shots."); 
    }
    
    public void PracticeTackling() 
    { 
        Console.WriteLine("  [Goalie] Practicing 1-on-1 box tackles."); 
    }
    
    public void PracticeDivingSaves() 
    { 
        Console.WriteLine("  [Goalie] Practicing top-corner diving saves."); 
    }
    
    public void PracticeHandDistribution() 
    { 
        Console.WriteLine("  [Goalie] Practicing fast break throws."); 
    }
}